### PR TITLE
104 add firmware module meta data field

### PR DESCRIPTION
**Pull request by weidenba**

### CODE REVIEW

File: src/objects/firmware.py [hunk 1]
```diff
@@ -54,10 +61,14 @@ def get_hid(self, root_uid=None):
         '''
         return a human readable identifier
         '''
-        return '{} {} - {}'.format(self.vendor, self.device_name, self.version)
+        if self.part == '':
```
Comment (dorpvom) [inline src/objects/firmware.py]: `part = ' - {}'.format(self.part) if self.part else ''`

File: src/storage/db_interface_frontend.py [hunk 2]
```diff
@@ -106,10 +110,18 @@ def _get_one_virtual_path_of_fo(fo_dict, root_uid):
...
     def _get_hid_firmware(self, uid):
-        firmware = self.firmwares.find_one({'_id': uid}, {'vendor': 1, 'device_name': 1, 'version': 1, 'device_class': 1})
+        firmware = self.firmwares.find_one({'_id': uid}, {'vendor': 1, 'device_name': 1, 'device_part': 1, 'version': 1, 'device_class': 1})
         if firmware is not None:
-            return '{} {} - {} ({})'.format(firmware['vendor'], firmware['device_name'], firmware['version'], firmware['device_class'])
-        return None
+            if 'device_part' not in firmware:  # for compatibility reasons
+                part = ' -'
+            else:
+                if firmware['device_part'] != '':
```
Comment (dorpvom) [inline src/storage/db_interface_frontend.py]: `part = ' - {}'.format(firmware['device_part']) if firmware['device_part'] != '' else ' -'`

or even

`part = ' -' if 'device_part' not in firmware or firmware['device_part'] == '' else ' - {}'.format(firmware['device_part'])`

File: src/test/unit/storage/test_db_interface_common.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import pytest
+
+from storage.db_interface_common import MongoInterfaceCommon
+
+
+class CommonInterfaceMock(MongoInterfaceCommon):
```
Comment (dorpvom) [inline src/test/unit/storage/test_db_interface_common.py]: YADM?

File: src/test/unit/storage/test_db_interface_common.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    'file_name': 'name_of_the_file',
+    'device_name': 'test_device',
+    'device_class': 'class_of_the_device',
+    'release_date': 0,
+    'vendor': 'test_vendor',
+    'version': '0.1',
+    'processed_analysis': {},
+    'files_included': [],
+    'virtual_file_path': {},
+    'comment': 'some comment'
```
Comment (dorpvom) [inline src/test/unit/storage/test_db_interface_common.py]: Where are comments in the new data format?

File: src/test/acceptance/test_analyze_firmware.py [hunk 5]
```diff
@@ -55,6 +55,7 @@ def _upload_firmware_post(self):
         with open(testfile_path, 'rb') as fp:
             data = {
                 'file': (fp, self.test_fw_a.file_name),
+                'device_part': 'full',
```
Comment (dorpvom) [inline src/test/acceptance/test_analyze_firmware.py]: care to check for this in result?

File: src/test/acceptance/rest/test_rest_analyze_firmware.py [hunk 6]
```diff
@@ -40,6 +40,7 @@ def _rest_upload_firmware(self):
             'binary': standard_b64encode(file_content).decode(),
             'file_name': 'test.zip',
             'device_name': 'test_device',
+            'device_part': 'full',
```
Comment (dorpvom) [inline src/test/acceptance/rest/test_rest_analyze_firmware.py]: care to check for this in result?

File: src/test/unit/web_interface/test_app_upload.py [hunk 7]
```diff
@@ -29,6 +30,7 @@ def test_app_upload_valid_firmware(self):
         rv = self.test_client.post('/upload', content_type='multipart/form-data', data={
             'file': (BytesIO(b'test_file_content'), 'test_file.txt'),
             'device_name': 'test_device',
+            'device_part': 'complete',
```
Comment (dorpvom) [inline src/test/unit/web_interface/test_app_upload.py]: care to check for this in result?

File: src/web_interface/templates/generic_view/firmware_detail_tabular_field.html [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<td class="active clickable" onclick="location.href='/analysis/{{ firmware[0] }}'">
+	{{ firmware[1] }}
+    {{ firmware[2]|render_tags(additional_class='pull-right')|safe}}
```
Comment (dorpvom) [inline src/web_interface/templates/generic_view/firmware_detail_tabular_field.html]: would look nicer if spaces are added around pipe operators

File: src/web_interface/templates/upload/re-analyze.html [hunk 9]
```diff
@@ -118,6 +118,20 @@ <h2>{{ title }} of {{ firmware.get_uid() | replace_uid_with_hid }}</h2>
                         </div>
                 </div>
 
+                <div class="form-group">
+                    <label class="control-label col-xs-3" for="device_part_dropdown">Device Part:</label>
+                        <div class="col-xs-9 {% if 'device_name' in error.keys() %}has-error{% endif %}">
```
Comment (dorpvom) [inline src/web_interface/templates/upload/re-analyze.html]: wrong indentation. same as all other dropdown blocks.
same in upload also.

File: src/test/acceptance/rest/test_rest_analyze_firmware.py (hunk 6 above)
Comment (weidenba) [reply src/test/acceptance/rest/test_rest_analyze_firmware.py]: tested in test/acceptance/rest/test_rest_analyze_firmware.py
+ BugFix in helperFunctions/object_conversion.py

File: src/test/acceptance/test_analyze_firmware.py (hunk 5 above)
Comment (weidenba) [reply src/test/acceptance/test_analyze_firmware.py]: tested in line 79

File: src/test/unit/storage/test_db_interface_common.py (hunk 4 above)
Comment (weidenba) [reply src/test/unit/storage/test_db_interface_common.py]: comments are stored in another collection in the new format

File: src/test/unit/web_interface/test_app_upload.py (hunk 7 above)
Comment (weidenba) [reply src/test/unit/web_interface/test_app_upload.py]: tested in acceptance test

**Pull request merged**